File: rpg/platform_app/tavern_cards.py
"""
tavern_cards.py — SillyTavern V1/V2 角色卡 import/export 兼容

支持：
- 导入 V1 (扁平 JSON) 和 V2 (spec_v2 + data 三层) 格式
- 导入 PNG 嵌入卡：解析 tEXt chunk 的 "chara" 关键字（V2 也用 "ccv3" / "chara"）
- 导出本人 user_character_cards / user_personas 为 V2 JSON

字段映射（V2 data → user_character_cards）：
  name              → name
  description       → identity
  personality       → personality
  scenario          → metadata.scenario
  first_mes         → metadata.first_mes
  mes_example       → 取首段对话进 sample_dialogue[0]
  creator_notes     → metadata.creator_notes（不入 prompt）
  system_prompt     → metadata.system_prompt
  alternate_greetings → metadata.alternate_greetings
  tags              → tags
  creator           → metadata.creator
  character_version → metadata.character_version
  extensions        → metadata.extensions
  character_book    → metadata.character_book（保留原结构，后续可接入世界书表）
"""
from __future__ import annotations

import base64
import binascii
import json
import re
import struct
import zlib
from typing import Any


# ── 解析 ──────────────────────────────────────────────────────────────
def parse_card(data: dict[str, Any] | str | bytes) -> dict[str, Any]:
    """统一入口：吃 dict / JSON 字符串 / base64 字符串，返回 V2 形态 dict。"""
    if isinstance(data, (bytes, bytearray)):
        text = data.decode("utf-8", errors="replace")
        return parse_card(text)
    if isinstance(data, str):
        # 可能是裸 JSON 或 base64
        stripped = data.strip()
        if stripped.startswith("{"):
            return parse_card(json.loads(stripped))
        try:
            decoded = base64.b64decode(stripped, validate=True).decode("utf-8")
            return parse_card(json.loads(decoded))
        except (binascii.Error, json.JSONDecodeError, UnicodeDecodeError) as exc:
            raise ValueError(f"无法解析角色卡：既不是 JSON 也不是 base64({exc})") from exc
    if not isinstance(data, dict):
        raise ValueError(f"不支持的角色卡类型：{type(data)}")
    # 是 V2 还是 V1？
    if data.get("spec") == "chara_card_v2" or data.get("spec") == "chara_card_v3":
        return _normalize_v2(data)
    return _v1_to_v2(data)


def _normalize_v2(card: dict[str, Any]) -> dict[str, Any]:
    """确保 V2 结构完整，补缺失字段。"""
    d = dict(card.get("data") or {})
    out = {
        "spec": card.get("spec") or "chara_card_v2",
        "spec_version": card.get("spec_version") or "2.0",
        "data": {
            "name": str(d.get("name") or "").strip(),
            "description": str(d.get("description") or ""),
            "personality": str(d.get("personality") or ""),
            "scenario": str(d.get("scenario") or ""),
            "first_mes": str(d.get("first_mes") or ""),
            "mes_example": str(d.get("mes_example") or ""),
            "creator_notes": str(d.get("creator_notes") or ""),
            "system_prompt": str(d.get("system_prompt") or ""),
            "post_history_instructions": str(d.get("post_history_instructions") or ""),
            "alternate_greetings": list(d.get("alternate_greetings") or []),
            "tags": list(d.get("tags") or []),
            "creator": str(d.get("creator") or ""),
            "character_version": str(d.get("character_version") or ""),
            "extensions": dict(d.get("extensions") or {}),
            "character_book": d.get("character_book"),
        },
    }
    if not out["data"]["name"]:
        raise ValueError("角色卡缺少 name")
    return out


def _v1_to_v2(card: dict[str, Any]) -> dict[str, Any]:
    """V1 扁平 → V2 标准化。"""
    name = (card.get("name") or card.get("char_name") or "").strip()
    if not name:
        raise ValueError("V1 角色卡缺少 name")
    return _normalize_v2({
        "spec": "chara_card_v1",
        "spec_version": "1.0",
        "data": {
            "name": name,
            "description": card.get("description", "") or card.get("char_persona", ""),
            "personality": card.get("personality", ""),
            "scenario": card.get("scenario", "") or card.get("world_scenario", ""),
            "first_mes": card.get("first_mes", "") or card.get("char_greeting", ""),
            "mes_example": card.get("mes_example", "") or card.get("example_dialogue", ""),
            "creator": card.get("creator", ""),
            "character_version": card.get("character_version", "1.0"),
            "tags": card.get("tags", []) or [],
        },
    })


# ── PNG tEXt chunk 解析 ──────────────────────────────────────────────
PNG_SIGNATURE = b"\x89PNG\r\n\x1a\n"


def parse_png_card(blob: bytes) -> dict[str, Any]:
    """从 PNG 文件读 tEXt/zTXt chunk 中的 chara 数据。"""
    if not blob.startswith(PNG_SIGNATURE):
        raise ValueError("不是合法 PNG 文件")
    offset = 8
    text_chunks: dict[str, str] = {}
    while offset < len(blob):
        if offset + 8 > len(blob):
            break
        length = struct.unpack(">I", blob[offset:offset + 4])[0]
        chunk_type = blob[offset + 4:offset + 8].decode("ascii", errors="replace")
        body = blob[offset + 8:offset + 8 + length]
        offset += 12 + length  # 4 type + length + 4 CRC
        if chunk_type == "IEND":
            break
        if chunk_type in ("tEXt", "zTXt"):
            try:
                if chunk_type == "tEXt":
                    key, _, value = body.partition(b"\x00")
                    text_chunks[key.decode("latin-1")] = value.decode("utf-8", errors="replace")
                else:  # zTXt：压缩文本
                    key, _, rest = body.partition(b"\x00")
                    # rest[0] 是 compression method（0=deflate），rest[1:] 是压缩数据
                    compressed = rest[1:] if len(rest) > 1 else b""
                    text_chunks[key.decode("latin-1")] = zlib.decompress(compressed).decode("utf-8", errors="replace")
            except Exception:
                continue
    # SillyTavern 通常用 key="chara" 或 "ccv3"
    for search_key in ("ccv3", "chara"):
        if search_key in text_chunks:
            return parse_card(text_chunks[search_key])
    raise ValueError("PNG 不包含 chara/ccv3 tEXt chunk")


# ── 映射到我方 user_character_cards ──────────────────────────────────
def tavern_to_user_card(card_v2: dict[str, Any]) -> dict[str, Any]:
    """V2 → user_character_cards.upsert_user_card() 的 payload。"""
    d = card_v2["data"]
    # mes_example 切第一条对话作为 sample_dialogue
    samples: list[str] = []
    for chunk in re.split(r"<START>|---", d.get("mes_example", "")):
        chunk = chunk.strip()
        if not chunk:
            continue
        # 提取 {{char}}: 后的内容
        for line in chunk.splitlines():
            line = line.strip()
            if not line:
                continue
            m = re.match(r"\{\{char\}\}:\s*(.+)", line)
            if m:
                samples.append(m.group(1).strip())
                if len(samples) >= 4:
                    break
        if samples:
            break

    return {
        "name": d["name"],
        "identity": d.get("description", "")[:2000],
        "personality": d.get("personality", "")[:1500],
        "speech_style": "",  # tavern 没单独字段，留空
        "current_status": "",
        "secrets": "",
        "sample_dialogue": samples,
        "tags": d.get("tags") or [],
        "metadata": {
            "tavern_imported": True,
            "scenario": d.get("scenario", ""),
            "first_mes": d.get("first_mes", ""),
            "alternate_greetings": d.get("alternate_greetings", []),
            "creator_notes": d.get("creator_notes", ""),
            "system_prompt": d.get("system_prompt", ""),
            "post_history_instructions": d.get("post_history_instructions", ""),
            "creator": d.get("creator", ""),
            "character_version": d.get("character_version", ""),
            "extensions": d.get("extensions") or {},
            "character_book": d.get("character_book"),
            "spec": card_v2.get("spec"),
            "spec_version": card_v2.get("spec_version"),
        },
    }


# ── 导出：user_character_cards → V2 JSON ─────────────────────────────
def write_png_card(v2_card: dict[str, Any], template_png: bytes | None = None) -> bytes:
    """把 V2 卡 JSON 嵌入 PNG 的 tEXt chara chunk。

    template_png: 可选 PNG 文件作底图；省略则生成一张 1x1 透明 PNG。
    """
    if template_png and template_png.startswith(PNG_SIGNATURE):
        png = template_png
    else:
        # 生成最小 1x1 透明 PNG
        png = _minimal_png()

    json_str = json.dumps(v2_card, ensure_ascii=False)
    chara_b64 = base64.b64encode(json_str.encode("utf-8"))
    chunk_data = b"chara" + b"\x00" + chara_b64
    text_chunk = (
        struct.pack(">I", len(chunk_data))
        + b"tEXt"
        + chunk_data
        + struct.pack(">I", zlib.crc32(b"tEXt" + chunk_data))
    )
    # 插到 IEND chunk 之前
    iend_pos = png.rfind(b"IEND")
    if iend_pos < 4:
        raise ValueError("template_png 没有 IEND chunk")
    # IEND chunk 起点（length 字段在 type 前 4 字节）
    insert_at = iend_pos - 4
    return png[:insert_at] + text_chunk + png[insert_at:]


def _minimal_png() -> bytes:
    """生成 1x1 透明 PNG，作为没传 template 时的默认底。"""
    sig = PNG_SIGNATURE
    # IHDR: 1x1, 8bit, RGBA
    ihdr_data = struct.pack(">IIBBBBB", 1, 1, 8, 6, 0, 0, 0)
    ihdr = struct.pack(">I", 13) + b"IHDR" + ihdr_data + struct.pack(">I", zlib.crc32(b"IHDR" + ihdr_data))
    # IDAT: 单像素透明（zlib 压缩 \x00 + 4 字节 RGBA）
    raw = b"\x00\x00\x00\x00\x00"  # filter byte + RGBA
    compressed = zlib.compress(raw)
    idat = struct.pack(">I", len(compressed)) + b"IDAT" + compressed + struct.pack(">I", zlib.crc32(b"IDAT" + compressed))
    iend = struct.pack(">I", 0) + b"IEND" + struct.pack(">I", zlib.crc32(b"IEND"))
    return sig + ihdr + idat + iend


def user_card_to_tavern_v2(card: dict[str, Any]) -> dict[str, Any]:
    """反向：本人卡 → V2 JSON 标准格式，可下载给酒馆用。"""
    md = card.get("metadata") or {}
    samples = card.get("sample_dialogue") or []
    # 合成 mes_example（SillyTavern 习惯）
    mes_example = ""
    if samples:
        sample_blocks = []
        for s in samples[:4]:
            sample_blocks.append(f"<START>\n{{{{user}}}}: \n{{{{char}}}}: {s}")
        mes_example = "\n".join(sample_blocks)

    return {
        "spec": "chara_card_v2",
        "spec_version": "2.0",
        "data": {
            "name": card.get("name", ""),
            "description": card.get("identity", "") or card.get("appearance", ""),
            "personality": card.get("personality", ""),
            "scenario": md.get("scenario", ""),
            "first_mes": md.get("first_mes", ""),
            "mes_example": md.get("mes_example") or mes_example,
            "creator_notes": md.get("creator_notes", ""),
            "system_prompt": md.get("system_prompt", ""),
            "post_history_instructions": md.get("post_history_instructions", ""),
            "alternate_greetings": md.get("alternate_greetings", []),
            "tags": card.get("tags") or [],
            "creator": md.get("creator", ""),
            "character_version": md.get("character_version", "1.0"),
            "extensions": md.get("extensions") or {},
            "character_book": md.get("character_book"),
        },
    }

File: rust/crates/rpg-platform/src/tavern_cards.rs
//! tavern_cards —— SillyTavern V1/V2 角色卡 import/export。
//!
//! 对应 Python: `rpg/platform_app/tavern_cards.py`。
//!
//! 支持:
//! - 解析 V1(扁平)/ V2(`spec=chara_card_v2`,`data` 三层)JSON。
//! - 解析 PNG `tEXt` / `zTXt` chunk 里的 `chara` / `ccv3` 关键字嵌入卡。
//! - 反向:把内部 user_card → V2 JSON、嵌回 PNG。
//!
//! 字段映射详见 `tavern_to_user_card`。

use base64::{engine::general_purpose, Engine};
use byteorder::{BigEndian, ByteOrder};
use flate2::read::ZlibDecoder;
use flate2::write::ZlibEncoder;
use flate2::Compression;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::io::{Read, Write};

use crate::error::{PlatformError, PlatformResult};

const PNG_SIGNATURE: [u8; 8] = [0x89, b'P', b'N', b'G', b'\r', b'\n', 0x1a, b'\n'];

/// V2 卡 wrapper(`{ spec, spec_version, data }`)。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TavernCard {
    pub spec: String,
    pub spec_version: String,
    pub data: TavernData,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TavernData {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub personality: String,
    #[serde(default)]
    pub scenario: String,
    #[serde(default)]
    pub first_mes: String,
    #[serde(default)]
    pub mes_example: String,
    #[serde(default)]
    pub creator_notes: String,
    #[serde(default)]
    pub system_prompt: String,
    #[serde(default)]
    pub post_history_instructions: String,
    #[serde(default)]
    pub alternate_greetings: Vec<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub creator: String,
    #[serde(default)]
    pub character_version: String,
    #[serde(default)]
    pub extensions: Value,
    #[serde(default)]
    pub character_book: Value,
}

fn s(v: Option<&Value>) -> String {
    v.and_then(|x| x.as_str()).map(|s| s.to_string()).unwrap_or_default()
}

fn list_of_str(v: Option<&Value>) -> Vec<String> {
    v.and_then(|x| x.as_array())
        .map(|arr| arr.iter().filter_map(|i| i.as_str().map(|s| s.to_string())).collect())
        .unwrap_or_default()
}

/// 入口:吃 JSON `Value` / 字符串 / 字节,返回标准化 V2。
pub fn parse_card_value(data: &Value) -> PlatformResult<TavernCard> {
    if !data.is_object() {
        return Err(PlatformError::validation(format!(
            "不支持的角色卡类型: {}",
            data
        )));
    }
    let spec = data.get("spec").and_then(|s| s.as_str()).unwrap_or("");
    if spec == "chara_card_v2" || spec == "chara_card_v3" {
        normalize_v2(data)
    } else {
        v1_to_v2(data)
    }
}

/// 从字符串解析(裸 JSON 或 base64(JSON))。
pub fn parse_card_str(text: &str) -> PlatformResult<TavernCard> {
    let stripped = text.trim();
    if stripped.starts_with('{') {
        let v: Value = serde_json::from_str(stripped)?;
        return parse_card_value(&v);
    }
    let decoded = general_purpose::STANDARD
        .decode(stripped.as_bytes())
        .map_err(|e| PlatformError::validation(format!("无法解析角色卡(非 JSON 也非 base64): {e}")))?;
    let text2 = String::from_utf8(decoded)
        .map_err(|e| PlatformError::validation(format!("无法解析角色卡(utf8): {e}")))?;
    let v: Value = serde_json::from_str(&text2)?;
    parse_card_value(&v)
}

fn normalize_v2(card: &Value) -> PlatformResult<TavernCard> {
    let d = card.get("data").cloned().unwrap_or(Value::Null);
    let do_obj = d.as_object();
    let name = s(do_obj.and_then(|o| o.get("name")));
    let name = name.trim().to_string();
    if name.is_empty() {
        return Err(PlatformError::validation("角色卡缺少 name"));
    }
    let spec = card
        .get("spec")
        .and_then(|s| s.as_str())
        .unwrap_or("chara_card_v2")
        .to_string();
    let spec_version = card
        .get("spec_version")
        .and_then(|s| s.as_str())
        .unwrap_or("2.0")
        .to_string();
    let data = TavernData {
        name,
        description: s(do_obj.and_then(|o| o.get("description"))),
        personality: s(do_obj.and_then(|o| o.get("personality"))),
        scenario: s(do_obj.and_then(|o| o.get("scenario"))),
        first_mes: s(do_obj.and_then(|o| o.get("first_mes"))),
        mes_example: s(do_obj.and_then(|o| o.get("mes_example"))),
        creator_notes: s(do_obj.and_then(|o| o.get("creator_notes"))),
        system_prompt: s(do_obj.and_then(|o| o.get("system_prompt"))),
        post_history_instructions: s(do_obj.and_then(|o| o.get("post_history_instructions"))),
        alternate_greetings: list_of_str(do_obj.and_then(|o| o.get("alternate_greetings"))),
        tags: list_of_str(do_obj.and_then(|o| o.get("tags"))),
        creator: s(do_obj.and_then(|o| o.get("creator"))),
        character_version: s(do_obj.and_then(|o| o.get("character_version"))),
        extensions: do_obj
            .and_then(|o| o.get("extensions"))
            .cloned()
            .unwrap_or(Value::Object(Default::default())),
        character_book: do_obj
            .and_then(|o| o.get("character_book"))
            .cloned()
            .unwrap_or(Value::Null),
    };
    Ok(TavernCard {
        spec,
        spec_version,
        data,
    })
}

fn v1_to_v2(card: &Value) -> PlatformResult<TavernCard> {
    let name_raw = s(card.get("name"));
    let name = if name_raw.is_empty() {
        s(card.get("char_name"))
    } else {
        name_raw
    };
    let name = name.trim().to_string();
    if name.is_empty() {
        return Err(PlatformError::validation("V1 角色卡缺少 name"));
    }
    let take = |a: &str, b: &str| -> String {
        let v = s(card.get(a));
        if v.is_empty() {
            s(card.get(b))
        } else {
            v
        }
    };
    let character_version = {
        let v = s(card.get("character_version"));
        if v.is_empty() { "1.0".to_string() } else { v }
    };
    let wrapped = json!({
        "spec": "chara_card_v1",
        "spec_version": "1.0",
        "data": {
            "name": name,
            "description": take("description", "char_persona"),
            "personality": s(card.get("personality")),
            "scenario": take("scenario", "world_scenario"),
            "first_mes": take("first_mes", "char_greeting"),
            "mes_example": take("mes_example", "example_dialogue"),
            "creator": s(card.get("creator")),
            "character_version": character_version,
            "tags": card.get("tags").cloned().unwrap_or(Value::Array(vec![])),
        }
    });
    normalize_v2(&wrapped)
}

// ─── PNG tEXt / zTXt 解析 ──────────────────────────────────────────────

/// 从 PNG 字节里读出嵌入的卡。
pub fn parse_png_card(blob: &[u8]) -> PlatformResult<TavernCard> {
    if blob.len() < 8 || blob[..8] != PNG_SIGNATURE {
        return Err(PlatformError::validation("不是合法 PNG 文件"));
    }
    let mut offset = 8usize;
    let mut text_chunks: HashMap<String, String> = HashMap::new();
    while offset + 8 <= blob.len() {
        let length = BigEndian::read_u32(&blob[offset..offset + 4]) as usize;
        let chunk_type = std::str::from_utf8(&blob[offset + 4..offset + 8])
            .unwrap_or("")
            .to_string();
        let body_start = offset + 8;
        if body_start + length + 4 > blob.len() {
            break;
        }
        let body = &blob[body_start..body_start + length];
        offset = body_start + length + 4; // skip CRC

        if chunk_type == "IEND" {
            break;
        }
        if chunk_type == "tEXt" {
            if let Some(idx) = body.iter().position(|&b| b == 0) {
                let key = String::from_utf8_lossy(&body[..idx]).into_owned();
                let val = String::from_utf8_lossy(&body[idx + 1..]).into_owned();
                text_chunks.insert(key, val);
            }
        } else if chunk_type == "zTXt" {
            if let Some(idx) = body.iter().position(|&b| b == 0) {
                let key = String::from_utf8_lossy(&body[..idx]).into_owned();
                // rest[0] = compression method, rest[1..] = compressed
                if idx + 2 <= body.len() {
                    let compressed = &body[idx + 2..];
                    let mut dec = ZlibDecoder::new(compressed);
                    let mut out = String::new();
                    if dec.read_to_string(&mut out).is_ok() {
                        text_chunks.insert(key, out);
                    }
                }
            }
        }
    }
    for key in ["ccv3", "chara"] {
        if let Some(raw) = text_chunks.get(key) {
            return parse_card_str(raw);
        }
    }
    Err(PlatformError::validation(
        "PNG 不包含 chara/ccv3 tEXt chunk",
    ))
}

// ─── Mapping: V2 ↔ user_character_cards payload ────────────────────────

/// V2 → `upsert_user_card` 用的 payload。返回 JSON `Value` 方便 user_cards 接收。
pub fn tavern_to_user_card(card: &TavernCard) -> Value {
    let d = &card.data;
    // 切 mes_example,提取首段 {{char}}: 后内容做 sample_dialogue。
    let mut samples: Vec<String> = Vec::new();
    'outer: for chunk in regex::Regex::new(r"<START>|---").unwrap().split(&d.mes_example) {
        let chunk = chunk.trim();
        if chunk.is_empty() {
            continue;
        }
        for line in chunk.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }
            if let Some(rest) = line.strip_prefix("{{char}}:") {
                samples.push(rest.trim().to_string());
                if samples.len() >= 4 {
                    break 'outer;
                }
            }
        }
        if !samples.is_empty() {
            break;
        }
    }

    let truncate = |s: &str, n: usize| -> String {
        s.chars().take(n).collect()
    };
    json!({
        "name": d.name,
        "identity": truncate(&d.description, 2000),
        "personality": truncate(&d.personality, 1500),
        "speech_style": "",
        "current_status": "",
        "secrets": "",
        "sample_dialogue": samples,
        "tags": d.tags,
        "metadata": {
            "tavern_imported": true,
            "scenario": d.scenario,
            "first_mes": d.first_mes,
            "alternate_greetings": d.alternate_greetings,
            "creator_notes": d.creator_notes,
            "system_prompt": d.system_prompt,
            "post_history_instructions": d.post_history_instructions,
            "creator": d.creator,
            "character_version": d.character_version,
            "extensions": d.extensions,
            "character_book": d.character_book,
            "spec": card.spec,
            "spec_version": card.spec_version,
        }
    })
}

/// 反向:本人卡 → V2 JSON。
pub fn user_card_to_tavern_v2(card: &Value) -> TavernCard {
    let md = card
        .get("metadata")
        .cloned()
        .unwrap_or(Value::Object(Default::default()));
    let samples = card
        .get("sample_dialogue")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();
    let mes_example = if !samples.is_empty() {
        samples
            .iter()
            .take(4)
            .filter_map(|s| s.as_str())
            .map(|s| format!("<START>\n{{{{user}}}}: \n{{{{char}}}}: {}", s))
            .collect::<Vec<_>>()
            .join("\n")
    } else {
        String::new()
    };
    let identity = {
        let v = s(card.get("identity"));
        if v.is_empty() { s(card.get("appearance")) } else { v }
    };
    let mes_example_md = s(md.get("mes_example"));
    let mes_example_final = if mes_example_md.is_empty() {
        mes_example
    } else {
        mes_example_md
    };
    TavernCard {
        spec: "chara_card_v2".to_string(),
        spec_version: "2.0".to_string(),
        data: TavernData {
            name: s(card.get("name")),
            description: identity,
            personality: s(card.get("personality")),
            scenario: s(md.get("scenario")),
            first_mes: s(md.get("first_mes")),
            mes_example: mes_example_final,
            creator_notes: s(md.get("creator_notes")),
            system_prompt: s(md.get("system_prompt")),
            post_history_instructions: s(md.get("post_history_instructions")),
            alternate_greetings: list_of_str(md.get("alternate_greetings")),
            tags: list_of_str(card.get("tags")),
            creator: s(md.get("creator")),
            character_version: {
                let v = s(md.get("character_version"));
                if v.is_empty() { "1.0".to_string() } else { v }
            },
            extensions: md
                .get("extensions")
                .cloned()
                .unwrap_or(Value::Object(Default::default())),
            character_book: md.get("character_book").cloned().unwrap_or(Value::Null),
        },
    }
}

// ─── PNG 嵌入导出 ──────────────────────────────────────────────────────

/// 把 V2 卡 JSON 嵌入 PNG 的 `tEXt chara` chunk。
pub fn write_png_card(card: &TavernCard, template_png: Option<&[u8]>) -> PlatformResult<Vec<u8>> {
    let png: Vec<u8> = match template_png {
        Some(t) if t.len() >= 8 && t[..8] == PNG_SIGNATURE => t.to_vec(),
        _ => minimal_png(),
    };

    let json_str = serde_json::to_string(card)?;
    let chara_b64 = general_purpose::STANDARD.encode(json_str.as_bytes());
    let mut chunk_data = Vec::new();
    chunk_data.extend_from_slice(b"chara");
    chunk_data.push(0);
    chunk_data.extend_from_slice(chara_b64.as_bytes());

    let mut text_chunk = Vec::new();
    let mut len_buf = [0u8; 4];
    BigEndian::write_u32(&mut len_buf, chunk_data.len() as u32);
    text_chunk.extend_from_slice(&len_buf);
    text_chunk.extend_from_slice(b"tEXt");
    text_chunk.extend_from_slice(&chunk_data);
    let mut hasher = crc32fast::Hasher::new();
    hasher.update(b"tEXt");
    hasher.update(&chunk_data);
    let crc = hasher.finalize();
    let mut crc_buf = [0u8; 4];
    BigEndian::write_u32(&mut crc_buf, crc);
    text_chunk.extend_from_slice(&crc_buf);

    // 在 IEND chunk 的 length 字段前插入。
    let iend_pos = png
        .windows(4)
        .rposition(|w| w == b"IEND")
        .ok_or_else(|| PlatformError::validation("template_png 没有 IEND chunk"))?;
    if iend_pos < 4 {
        return Err(PlatformError::validation("template_png IEND 偏移异常"));
    }
    let insert_at = iend_pos - 4;
    let mut out = Vec::with_capacity(png.len() + text_chunk.len());
    out.extend_from_slice(&png[..insert_at]);
    out.extend_from_slice(&text_chunk);
    out.extend_from_slice(&png[insert_at..]);
    Ok(out)
}

fn minimal_png() -> Vec<u8> {
    // 1x1 透明 RGBA。
    let mut out = Vec::new();
    out.extend_from_slice(&PNG_SIGNATURE);

    // IHDR
    let mut ihdr_data = Vec::with_capacity(13);
    let mut buf = [0u8; 4];
    BigEndian::write_u32(&mut buf, 1);
    ihdr_data.extend_from_slice(&buf);
    ihdr_data.extend_from_slice(&buf);
    ihdr_data.extend_from_slice(&[8, 6, 0, 0, 0]);
    push_chunk(&mut out, b"IHDR", &ihdr_data);

    // IDAT (zlib(filter byte + RGBA))
    let raw = [0u8; 5];
    let mut enc = ZlibEncoder::new(Vec::new(), Compression::default());
    enc.write_all(&raw).expect("zlib raw");
    let compressed = enc.finish().expect("zlib finish");
    push_chunk(&mut out, b"IDAT", &compressed);

    // IEND
    push_chunk(&mut out, b"IEND", &[]);
    out
}

fn push_chunk(out: &mut Vec<u8>, ty: &[u8; 4], data: &[u8]) {
    let mut len_buf = [0u8; 4];
    BigEndian::write_u32(&mut len_buf, data.len() as u32);
    out.extend_from_slice(&len_buf);
    out.extend_from_slice(ty);
    out.extend_from_slice(data);
    let mut hasher = crc32fast::Hasher::new();
    hasher.update(ty);
    hasher.update(data);
    let crc = hasher.finalize();
    let mut crc_buf = [0u8; 4];
    BigEndian::write_u32(&mut crc_buf, crc);
    out.extend_from_slice(&crc_buf);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn v1_minimal() {
        let v = json!({"name": "Alice", "description": "test"});
        let c = parse_card_value(&v).unwrap();
        assert_eq!(c.data.name, "Alice");
        assert_eq!(c.spec, "chara_card_v1");
    }

    #[test]
    fn png_roundtrip() {
        let card = TavernCard {
            spec: "chara_card_v2".to_string(),
            spec_version: "2.0".to_string(),
            data: TavernData { name: "Bob".to_string(), ..Default::default() },
        };
        let blob = write_png_card(&card, None).unwrap();
        let back = parse_png_card(&blob).unwrap();
        assert_eq!(back.data.name, "Bob");
    }
}
